### ISSUE TITLE
Convert Carbon-Bot to Rasa 2.x

### ISSUE THREAD

**Issue by kedz** [RasaHQ/carbon-bot#38]
### Goal
Get CB working in `rasa shell` under Rasa 2.x. The new version should be as faithful to original as possible because we want to benchmark any subsequent changes. This is not completely possible however as some policies that were in use are now deprecated.

Updates happening on `phoenix-bot` branch.

### Tasks to complete goal
- [x] Run `rasa data convert nlu|nlg|core -f yaml --data={SOURCE_DIR} --out={TARGET_DIR}`
- [x] Run `rasa train` and record warnings and errors. 
- [x] Fix missing conveRT url exception
- [x] Convert unfeaturized slot types to ~~`influence_conversation: false`~~ `any`
- [x] Move responses to domain file.
- [x] Update forms to 2.0
- [x] Add action_calculate_offsets actions to train/test stories
- [ ] ~~(Optionally) FIx stopover slot filling in `airtravel_form`.~~ decided against this for now.
- [x] Fix entities that are not slots but treated like slots.
- [x] Convert test stories to 2.0 format/sort out why they make the evaluation crash.
- [x] Clean up and organize training stories. ~Figure out why there are duplicate slot setting events.~
- [ ] ~~Use test story to performance to determine if the bot is working sufficiently similar to its original behavior.~~ bot performs well on the test set, and seems to behave similarly to original behaviour... but the test set is small. More rigorous investigation of this would probably mean creating a new test set by digging through old rasa x conversations
- [x] Update Rasa and Rasa X versions in the requirements data.

**Comment by kedz** [comment]
After converting the 1.x training data and running `rasa train` yields 2 categories of warnings
- Warnings that responses are not in domain file
- Unfeaturized slot is deprecated in favor of influence_conversation flag.

and an exception:
 - No conveRT url leads to training failure

**Comment by kedz** [comment]
After fixing the above issues I ran into a couple of other problems. 
- First some of the policies that the original bot used are no longer supported. 
  - Mapping and Form policies should be RulePolicy.
  - EmbeddingPolicy should be changed to TEDPolicy
  - TEDPolicy needs to train for many more epochs before it behaves reasonably, increasing from 50 to 300.
- Even with increased training time, the bot doesn't seem to follow expected story behavior and won't activate the form to get info.
- Some entities are referenced like slots but never defined as such. Hard to tell if that is making a difference.
- Some of training stories have weird duplicate slot setting actions that I don't understand
- `rasa test` on converted test stories crashes some reason related to trying to compare NoneTypes with >. Note sure what is happening yet. 
- Test stories also appear to be e2e stories  according to `rasa test` but don't appear to be so upon inspection. Not sure what is happening there.
- The main form is defined in a 1.0 style way and needs to be converted manually to 2.0.

**Comment by kedz** [comment]
Removed the city slot settings in test stories (the departure/arrival slots are still set), and removed entities labels on intents, which I believe is intended only to be used for e2e stories. This fixes the test stories and now they run without crashing.

**Comment by samsucik** [comment]
@kedz I've seen duplicated `SlotSet` events being a result of form validation: the slot gets set once in the form action (i.e. in the form) and then again when it gets validated. Might be the case here as well, and I'd just ignore it. If it happens even today, then we can fix it relatively easily I think.

**Comment by kedz** [comment]
Thanks for the heads up @samsucik!

**Comment by kedz** [comment]
Converted `airtravel_form` to 2.0 format. Handling of stopover cities seems broken now (although I can't tell if it worked in the 1.0 version either).  Also since there is no longer a `submit` method on form completion I had to introduce a custom action `action_calculate_offsets` that performs that logic. This action will need to be added to the train and test stories after the `active_loop: null` step.

I could try to fix the stopover behavior. I think it might be possible with some dynamic form behavior, but I might also need to add more nlu training examples since this is not at all well covered currently. I don't really want to do any improvements until we are actually in a position to run a quantitative evaluation though.

**Comment by koernerfelicia** [comment]
Really baffled by the bot's poor performance. Here's a failed test story (truncated):
```
- story: Story from conversation with b59c8f85-caea-4143-bae0-62421e93b664 on October 27th 2019 (/Users/felicia/Repositories/carbon-bot/test_stories/stories.yml)
  steps:
  - intent: greet
  - action: utter_SCENARIOCHECK
  - intent: SCENARIO
  - slot_was_set:
    - context_scenario: holiday
  - slot_was_set:
    - holiday_name: thanksgiving
  - action: action_start  # predicted: action_default_fallback
```

here's a story *from the training data*:

```
- story: give_travel_plan (2)
  steps:
  - intent: greet
  - action: utter_SCENARIOCHECK
  - intent: SCENARIO
    entities:
    - context_scenario: holiday
    - holiday_name: thanksgiving
  - slot_was_set:
    - context_scenario: holiday
  - slot_was_set:
    - holiday_name: thanksgiving
  - action: action_start
  ```
We have `MemoizationPolicy` in the configuration... so why wouldn't this work?

**Comment by koernerfelicia** [comment]
Update on this:

`MemoizationPolicy` wasn't being applied because there were too many conflicting stories. This appears to be a result of `action: airtravel_form` being peppered through the `airtravel_form` loop, which doesn't gel with how we do forms now.

I attempted to replace these with explicit slot-filling behaviour, but this lead to warnings about conflicts between rules and stories, where the rule was really just the loop rule from the `airtravel_form`. So, I removed the slot-filling behaviour within the stories (this should be covered by the form anyways), and `MemoizationPolicy` was applied properly, meaning test on the train stories results in 100% accuracy.

The one behaviour I wasn't able to preserve (that I know of...) was this pattern:
```
  - slot_was_set:
    - requested_slot: travel_flight_class
  - intent: why
  - action: utter_explain_economy_class
```
I could turn this into a rule, though in the test story it appears in this form:
```
  - slot_was_set:
    - requested_slot: travel_flight_class
  - intent: affirm
  - slot_was_set:
    - requested_slot: travel_flight_class
  - intent: why
  - action: utter_explain_economy_class
```
Not quite sure how to bake this into a rule, as it would need to be conditioned not only on the intent (why) but also on the action preceding it. Or maybe it could be conditioned on the `requested_slot`. Not sure, because the test story seems to be broken: in the original train stories `affirm` filled the `travel_flight_class` with `economy`,, in this test story it doesn't seem to fill the slot. Will have a think and come back to this after my self-care days.

**Comment by samsucik** [comment]
@koernerfelicia I think it's safe to ignore that test story, which clearly didn't go well and should be thrown away or corrected.

I think that in the light of [this fix](https://github.com/RasaHQ/rasa/pull/8628) it should be safe to write a rule like this, or not? Maybe I'm missing something...
```yml
steps:
- action: airtravel_form
- slot_was_set:
  - requested_slot: travel_flight_class
- intent: why
- action: utter_explain_economy_class
- action: airtravel_form
```

Edit: I'm unsure if this should be added to the end of the rule; at least the repeated `SlotSet` could probably be omitted...:
```yml
- active_loop: airtravel_form
- slot_was_set:
  - requested_slot: travel_flight_class
```

**Comment by koernerfelicia** [comment]
Thanks, @samsucik. I think if we decide to ignore that behaviour in the stories that rule works just great! I don't **think** that last bit needs to be added (at least it seems to be working fine).

Am now looking at the recovery for failed slot-filling. It seems that the bot gets stuck like so:

```
Will you be travelling in Economy Class?
Your input ->  yes 
What city will you be flying from?
Your input ->  Monich 
Your input ->       
```

**Comment by koernerfelicia** [comment]
Sam kindly talked the above through with me. Looks like this happens when no entities at all are picked up, so the form is rejected:

```
2021-05-14 16:50:14 DEBUG   rasa.core.policies.rule_policy - There is no applicable rule.
2021-05-14 16:50:14 DEBUG   rasa.core.policies.ensemble - Execution of 'airtravel_form' was rejected. Setting its confidence to 0.0 in all predictions.
2021-05-14 16:50:14 DEBUG   rasa.core.policies.ensemble - Made prediction using user intent.
2021-05-14 16:50:14 DEBUG   rasa.core.policies.ensemble - Added `DefinePrevUserUtteredFeaturization(False)` event.
2021-05-14 16:50:14 DEBUG   rasa.core.policies.ensemble - Predicted next action using policy_1_TEDPolicy.
2021-05-14 16:50:14 DEBUG   rasa.core.tracker_store - Recreating tracker for id 'ba806bb396d9459aae4f220535b1f372'
```
We think this used to work better because stories explicitly included form-type actions, roughly like so:
```
slot_was_set:
- requested_slot: departure
action: airtravel_form
intent: inform
slot_was_set:
- city: new york
```
And I replaced all of these with:
```
action: airtravel_form
active_loop: airtravel_form
active_loop: null
```
So, back to the drawing board, will try and revert some of the simplifications I made.

**Comment by samsucik** [comment]
As I'm giving it an N-th thought now, I think that we'll still have to come up with something else to solve the issue itself... The form rejection that we discussed happens when the intent is other than `inform`. If we add a story that includes a rejection caused by an entity-less `intent: inform`, followed by `action: airtravel_form`, then the non-rule policies can learn to bring us back on track in such a situation. But the ensemble will still fall back for rejections where the intent is different from `inform`. A few things to try might be:
- disable Core fallback
- write rules for specific rejections that we wanna handle (e.g. FAQ). in this sense, writing the full stories with form flows is kinds useless -- we can just write a rule that will handle an entity-less `inform` and bring the form back on track

My guess is that disabling fallback would bring us to where Carbon was pre-2.0 -- if I'm honest, I think that the old Carbon would've also failed whenever an entity-less intent would appear that wouldn't be `inform` or some covered intent like FAQ...

**Comment by koernerfelicia** [comment]
That's a very good point. I think I'll focus on reverting some of the changes and then see what that + disabling fallback does for us.

**Comment by koernerfelicia** [comment]
I've reverted the relevant changes (as yet unpushed), and am now seeing weird behaviour from MemoizationPolicy again, though this time it doesn't directly relate to form actions. Something is wonky about the prediction of `utter_explain_why_offset_travel`, TED kicks in below despite the existence of this exact pattern in training stories:
```
- story: manual_story_2 (/Users/felicia/Repositories/carbon-bot/test_stories.yml)
  steps:
  - intent: greet
  - action: utter_SCENARIOCHECK
  - intent: SCENARIO
  - slot_was_set:
    - context_scenario: holiday
  - slot_was_set:
    - holiday_name: thanksgiving
  - action: action_start
  - action: action_disclaimer
  - action: utter_holiday-travel_offer_help
  - intent: affirm
  - action: utter_explain_why_offset_travel  # predicted: utter_acknowledge
```
  The above story is from training, but just for kicks here's another training story with that same pattern. 
  
  ```
  - story: interactive_story_2
  steps:
  - intent: greet
  - action: utter_SCENARIOCHECK
  - intent: SCENARIO
  - slot_was_set:
    - context_scenario: holiday
  - slot_was_set:
    - holiday_name: thanksgiving
  - action: action_start
  - action: action_disclaimer
  - action: utter_holiday-travel_offer_help
  - intent: affirm
  - action: utter_explain_why_offset_travel
  ```
  
There doesn't seem to be a conflicting pattern, this pattern occurs 8 times in the training data. Am trying to dig through the MemoizationPolicy lookup to understand why..

**Comment by koernerfelicia** [comment]
MemoizationPolicy now seems to be working as expected. It seems that there were some conflicts that aren't picked up by story structure or rule validation: these all seemed to be either
`action: airtravel_form` within the `airtravel_form` loop or
incorrect order of intents and slot-setting events within the `airtravel_form` loop.

Upon fixing these, test performance on train and test data went up to 100% and nearly 100% respectively. `rasa shell`, however, failed (by which I mean, couldn't handle the happy path which it had essentially seen before).

This turned out to be due to the missing entities, which were present  in the prediction source when running shell, but had been removed from training stories so that I could run `rasa test core` on the training stories. `rasa shell` now seems to behave as expected (meaning, happy paths and mostly happy paths in which misspelled entities are supplied. I have not tested more rigorously).

Finally, I also attempted to address the following scenario, which old carbon bot could handle (at least, it seems this way from looking at rasa x conversations), but phoenix bot had trouble with:

user supplies a misspelled entity *that isn't picked up as an entity* and is instead assigned an intent, throwing core off. After discussing with Sam, we thought that adding the user interactions with the form back in would help with this:
before:
```
  - action: airtravel_form
  - active_loop: airtravel_form
  - active_loop: null
  - slot_was_set:
    - requested_slot: null
```
after:
```
  - action: airtravel_form
  - active_loop: airtravel_form
  - slot_was_set:
    - requested_slot: travel_flight_class
  - intent: affirm
  - slot_was_set:
    - travel_flight_class: economy
  - slot_was_set:
    - requested_slot: travel_departure
  - intent: inform
  - slot_was_set:
    - travel_departure: Berlin
  - slot_was_set:
    - requested_slot: travel_destination
  - intent: inform
  - slot_was_set:
    - travel_destination: Dublin
  - active_loop: null
  - slot_was_set:
    - requested_slot: null
```
But it didn't seem to. To be fair, we expected the above to only help if the utterance was assigned `inform`.

Instead, I added a `from_text` to the form for both departure and destination city, so that a misspelled entity that is not picked up by DIET can still be rejected by the form validation, causing the bot to re-ask. This seems to work.

I still don't know why the old carbon bot handled this better. I imagine it's some difference in old handling of forms.

**Comment by koernerfelicia** [comment]
I also took the liberty of removing stopover and reference resolution code and data. Just to keep everything cleaner. If we want to expand Carbon-bot's features to include stopover and reference resolution we'll have to add these things back in.
